### PR TITLE
Refactor several distributions

### PR DESCRIPTION
**Pull request by ricardoV94**
Couple more of distributions refactored to `v4`. This should cover all the ones that are to be implemented on Aesara, unless I missed something.

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes?
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [x] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: @brandonwillard Is there a more elegant solution? The problem was that `BoundedContinuous` is looking for the `lower/upper` inputs based on order, and in this case it's the first and last inputs that it should use.

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]:
```suggestion
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [inline pymc3/distributions/continuous.py:2653]: Instead of creating a class that's just a constant times an existing `RandomVariable`, let's just create the product in `Distribution.dist`.

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [reply pymc3/distributions/continuous.py]: Yeah, we don't want to add properties to objects like this.  As a matter of fact, `Op`s are likely to use `__slots__` in the near future, making this an error.

**Comment by ricardoV94** [comment]
@brandonwillard What do you think of my last commit as a solution for the BoundContinuous issue? 

I moved the `transform_logp` inside `pm.Distribution.__new__()` instead of being inside `model.create_value_var()`. This allows to pass class attributes as optional arguments to `transform_logp` without cluttering the rv_op. 

BTW I see this was before in the metaclass: https://github.com/pymc-devs/pymc3/blob/1e4163339f963d9728d0160e74d594f685d76b92/pymc3/distributions/distribution.py#L112-L120

Should we move it there or remove those commented lines?

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:2653]: How will we pass the beta argument to the logp/logcdf?

Comment (brandonwillard) [reply pymc3/distributions/continuous.py:2653]: Since this is just a scaled random variable, we would need to write the `logpt` dispatch that formulates the log-likelihood for a scale random variable.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py:2653]: That goes towards the idea of having that automatic logp inference for continuous transformations.

It's a great functionality to add but I think it's deserves it's own PR. Some work has to be done to figure out how it would fit with the automatic variable registration and automatic transformations inside the model object.

I am happy to either remove this distribution from the PR or add a clear TODO comment above it for now.

Comment (brandonwillard) [reply pymc3/distributions/continuous.py:2653]: I'm guessing it can be done in a single commit, so it might as well go here&mdash;or any other PR that needs it.

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: You could just override `Distribution.__new__` and create the transform (see [`Dirchlet.__new__`](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/multivariate.py#L390)).

**Review by brandonwillard — CHANGES_REQUESTED** [review]
We don't want to start adding variables at the class level unnecessarily; instead, each class that wants to use an `interval` transform should just create its own in `Distribution.__new__`.  That approach is class-level enough to accomplish the same thing, and it makes the transform-setup logic a little less spread out.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: In that case do we want to keep the bounded continuous class for anything? Maybe to raise an error if transform is UNSET?

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:2653]: Will take care of here: https://github.com/pymc-devs/pymc3/pull/4653

File: pymc3/distributions/distribution.py
Comment (twiecki) [inline pymc3/distributions/distribution.py:112]: Should just remove this.

File: pymc3/tests/test_distributions_random.py
Comment (twiecki) [inline pymc3/tests/test_distributions_random.py]: Shouldn't we just remove it then?

File: pymc3/distributions/distribution.py
Comment (ricardoV94) [reply pymc3/distributions/distribution.py:112]: Done

File: pymc3/tests/test_distributions_random.py
Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: Done

**Review by twiecki — DISMISSED** [review]

**Comment by twiecki** [comment]
```

=================================== FAILURES ===================================
_______________ TestMatchesScipy.test_beta_binomial_distribution _______________

self = <pymc3.tests.test_distributions.TestMatchesScipy object at 0x12affbeb0>

    @pytest.mark.xfail(condition=(aesara.config.floatX == "float32"), reason="Fails on float32")
    def test_beta_binomial_distribution(self):
>       self.checkd(
            BetaBinomial,
            Nat,
            {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
        )

pymc3/tests/test_distributions.py:1501: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pymc3.tests.test_distributions.TestMatchesScipy object at 0x12affbeb0>
distfam = <class 'pymc3.distributions.discrete.BetaBinomial'>
valuedomain = <pymc3.tests.test_distributions.Domain object at 0x12511d880>
vardomains = {'alpha': <pymc3.tests.test_distributions.Domain object at 0x12560c7f0>, 'beta': <pymc3.tests.test_distributions.Domain object at 0x12560c7f0>, 'n': <pymc3.tests.test_distributions.Domain object at 0x12511dfd0>}
checks = (<bound method TestMatchesScipy.check_int_to_1 of <pymc3.tests.test_distributions.TestMatchesScipy object at 0x12affbeb0>>,)
extra_args = {}

    def checkd(self, distfam, valuedomain, vardomains, checks=None, extra_args=None):
        if checks is None:
            checks = (self.check_int_to_1,)
    
        if extra_args is None:
            extra_args = {}
        m = build_model(distfam, valuedomain, vardomains, extra_args=extra_args)
        for check in checks:
>           check(m, m.named_vars["value"], valuedomain, vardomains)
E           AttributeError: 'tuple' object has no attribute 'named_vars'

pymc3/tests/test_distributions.py:863: AttributeError
=============================== warnings summary ===============================
```

**Comment by ricardoV94** [comment]
Yeah, it's using this `checkd` test that has not been refactored for V4. Just marked as xfail for now. Will fix later when refactoring the ZeroInflated distributions that make use of it as well.

**Comment by ricardoV94** [comment]
I added a skipif for the betabinomial test in case we are using scipy < 1.4.0 (which happens in the pytest workflow). The tests should still be covered in the arviz-compat workflow so it should be fine for now.

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [inline pymc3/distributions/continuous.py]: Since this field is very specific to upper and lower bounds, its name should reflect that.  Currently, it seems like this field specifies which arguments are to be transformed, which is very misleading.  Also, the values in this field are argument _indices_, so that should be made clearer somehow (e.g. `upper_lower_bound_indices`, `trans_ul_bound_indices`, etc.).

File: pymc3/distributions/continuous.py
Comment (brandonwillard) [inline pymc3/distributions/continuous.py:2647]: Let's create an issue specifically for this before/after merging.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Aside from renaming that transform field, it looks like `test_distributions_random` could be introducing a significant amount of testing redundancy.  Regardless, we can merge this once the transform field name is updated.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Renamed to `bound_args_indices`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:2647]: Done: https://github.com/pymc-devs/pymc3/issues/4683

File: pymc3/distributions/continuous.py
Comment (twiecki) [inline pymc3/distributions/continuous.py]: Why `2`? And shouldn't these be tuples instead of lists?

File: pymc3/distributions/continuous.py
Comment (twiecki) [inline pymc3/distributions/continuous.py:3744]: Why don't we need `self` here?

**Review by twiecki — DISMISSED** [review]

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Changed to tuples. 

The arguments of the triangular RandomOp are in this order `(lower, mode, upper)`, and so we need to specify that the first and last are the ones relevant for the transformation, hence `(0, 2)` (zero-based index)

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:3744]: Because these are used only as a sort of static-method by the `_logp` dispatcher here: https://github.com/pymc-devs/pymc3/blob/6c247638c506e4b3d6eff86ff24c5e98d34ae055/pymc3/distributions/distribution.py#L96-L102

We are not really using distribution instances anymore IIUC.

Comment (twiecki) [reply pymc3/distributions/continuous.py:3744]: What a curious pattern, thanks for explaining.

**Comment by twiecki** [comment]
This gets us a big step closer to merging `v4`, thanks @ricardoV94!

**Pull request merged**
